src/app/qgsgeometryvalidationservice.cpp

```diff
@@ -0,0 +1,454 @@
+  //  2. and looking it up from the edit buffer (in memory) is really fast.
+  // so in short: it's still a good idea, but not as important as on first thought.
+
+  if ( !mLayerChecks[layer].topologyChecks.empty() )
+  {
+    invalidateTopologyChecks( layer );
+  }
+
+  processFeature( layer, fid );
+}
+
+void QgsGeometryValidationService::onFeatureDeleted( QgsVectorLayer *layer, QgsFeatureId fid )
+{
+  if ( !mLayerChecks[layer].topologyChecks.empty() )
+  {
+    invalidateTopologyChecks( layer );
+  }
+
+  mLayerChecks[layer].singleFeatureCheckErrors.remove( fid );
+
+  // There should be no geometry errors on an inexistent feature, right?
+  emit geometryCheckCompleted( layer, fid, QList<std::shared_ptr<QgsSingleGeometryCheckError>>() );
+}
+
+void QgsGeometryValidationService::onBeforeCommitChanges( QgsVectorLayer *layer )
+{
+  if ( !mLayerChecks[layer].topologyChecks.empty() ) // TODO && topologyChecks not fulfilled
+  {
+    if ( !layer->allowCommit() )
+    {
+      showMessage( tr( "Running geometry validation checks before saving ..." ) );
+    }
+
+    mLayerChecks[layer].commitPending = true;
+
+    triggerTopologyChecks( layer );
+  }
+}
+
+void QgsGeometryValidationService::onEditingStopped( QgsVectorLayer *layer )
+{
+  cancelTopologyCheck( layer );
+  clearTopologyChecks( layer );
+}
+
+void QgsGeometryValidationService::showMessage( const QString &message )
+{
+  mMessageBar->popWidget( mMessageBarItem );
+  mMessageBarItem = QgsMessageBar::createMessage( tr( "Geometry Validation" ),  message );
+  mMessageBarItem->setDuration( 5 );
+  mMessageBar->pushItem( mMessageBarItem );
+}
+
+void QgsGeometryValidationService::cleanupLayerChecks( QgsVectorLayer *layer )
+{
+  if ( !mLayerChecks.contains( layer ) )
+    return;
+
+  VectorLayerCheckInformation &checkInformation = mLayerChecks[layer];
+
+  cancelTopologyCheck( layer );
+
+  qDeleteAll( checkInformation.singleFeatureChecks );
+  qDeleteAll( checkInformation.topologyChecks );
+  checkInformation.context.reset();
+}
+
+void QgsGeometryValidationService::enableLayerChecks( QgsVectorLayer *layer )
+{
+  if ( layer->geometryOptions()->geometryChecks().empty() && !mLayerChecks.contains( layer ) )
+    return;
+
+  VectorLayerCheckInformation &checkInformation = mLayerChecks[layer];
+
+  cleanupLayerChecks( layer );
+
+  if ( layer->geometryOptions()->geometryChecks().empty() )
+  {
+    for ( QMetaObject::Connection connection : qgis::as_const( checkInformation.connections ) )
+    {
+      disconnect( connection );
+    }
+    checkInformation.connections.clear();
+    return;
+  }
+
+  int precision = 0;
+  if ( layer->geometryOptions()->geometryPrecision() == 0 )
+    precision = 8;
+  else
+  {
+    precision = log10( layer->geometryOptions()->geometryPrecision() ) * -1;
+
+    if ( precision < 1 || precision > 13 )
+      precision = 8;
+  }
+
+  checkInformation.context = qgis::make_unique<QgsGeometryCheckContext>( 10, mProject->crs(), mProject->transformContext() );
```

Comment:
should 10 be `precision` here? (not fully sure on the details of these classes)

Reply:
Yep, this value might still need to be fine tuned